Deletion of a stop from pattern stops

br648 opened the pull request:
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [x] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

Added a new end point which allows a stop to be deleted along with all references in stop times and pattern stops. Once deleted the altered pattern stops are normalized to address the removed stop.

Jira: https://ibisolutions.atlassian.net/browse/DT-322

Commit:
feat(Allow the deletion of a stop from pattern stops): Updated EditorController to include a new end

philip-cline commented on src/main/java/com/conveyal/datatools/editor/controllers/api/EditorController.java:
```diff
@@ -258,6 +265,65 @@ private String deleteTripsForPattern(Request req, Response res) {
+        try (
+            Connection connection = datasource.getConnection();
+            PreparedStatement statement = connection.prepareStatement(
+                String.format("select pattern_id, stop_sequence from %s.pattern_stops where stop_id = ?", namespace)
```
Would it be cleaner to format with the `stopId` here?

br648 replied:
Sorry I don't follow, can you elaborate please?

philip-cline replied:
Oh I meant like `String.format("select id, stop_sequence from %s.pattern_stops where %s = %s", namespace, stopIdColumnName, stopId)` so using the String format rather than the PreparedStatement parameter

philip-cline replied:
I'm not really sure it's cleaner, just a thought since we're not really reusing the PreparedStatement

br648 replied:
I happened to address this in a follow up commit, so all good.

philip-cline commented on src/main/java/com/conveyal/datatools/editor/controllers/api/EditorController.java:
```diff
@@ -258,6 +265,65 @@ private String deleteTripsForPattern(Request req, Response res) {
+            Map<Integer, Integer> patternsToBeNormalized = new HashMap<>();
+            while (resultSet.next()) {
+                patternsToBeNormalized.put(
+                    resultSet.getInt("pattern_id"),
```
I don't think we're guaranteed that the `patternId` is an `int`: 
<img width="279" alt="image" src="https://github.com/ibi-group/datatools-server/assets/63798641/a473adec-5ca5-4451-9b62-11cdc60659e9">

br648 replied:
Good spot. I should be using the id and not pattern_id (they are the same in the test data). Fixed.

philip-cline replied:
Yeah it's tricky. that got me too with the other pattern stop PR

philip-cline reviewed (changes requested):
Looking really good for pattern_stops, but have a few comments and we'll need to remove the stop from stop_time entries as well.

Commit:
refactor(Updated scope): Now deleted stop and related stop times

philip-cline committed:
Merge branch 'dev' into feature/DT-322-Delete-stops-from-pattern-stops

philip-cline commented:
Merged dev just so we could get the pattern id updates for testing

philip-cline commented on src/main/java/com/conveyal/datatools/editor/controllers/api/EditorController.java:
```diff
@@ -129,6 +132,10 @@ private void registerRoutes() {
         }
+
+        if ("stop".equals(classToLowercase)) {
+            patch(ROOT_ROUTE + "/cascadeDeleteStop", this::cascadeDeleteStop, json::write);
```
Could we make the endpoint a `delete` or just a `get` endpoint? `patch` requests have some annoying requirements that make things a bit difficult on the front end

br648 replied:
@philip-cline I've changed to a delete. Note that the stop id now has to be part of the URL and not a query param:
`/api/editor/secure/stop/STOP_ID_HERE/cascadeDeleteStop?feedId=%s&sessionId=test`

philip-cline commented on src/test/java/com/conveyal/datatools/editor/controllers/api/EditorControllerTest.java:
```diff
@@ -143,6 +147,60 @@ public void canPatchStopsConditionally() throws IOException {
+        );
+        String response = given()
+            .port(DataManager.PORT)
+            .patch(path)
```
Just marking that this would have to be updated if we change the endpoint type

br648 replied:
Noted and updated to delete.

philip-cline reviewed (approved):
Testing this from some rough front end code looked good. Thanks for all the changes.

Commit:
refactor(Addressed PR feedback): End point type is now delete

Commit:
Merge branch 'feature/DT-322-Delete-stops-from-pattern-stops' of https://github.com/ibi-group/datatools-server into feature/DT-322-Delete-stops-from-pattern-stops

Commit:
refactor(EditorControllerTest.java): Corrected formatting.

Commit:
refactor(EditorControllerTest.java): Created a distinct feed version for the cascade delete stop tes

Deleting a stop from the previously shared feed version was causing another test to fail.

miles-grant-ibigroup reviewed (approved):
Working well and code looks clean thanks for these changes!
